### ISSUE TITLE
[Android]For the first time on android, there's no sound.

### ISSUE THREAD

**Issue by whitecostume** [libgdx/libgdx#5051]
#### Issue details
For the first time on android, there was no sound, only the second time.

#### Reproduction steps/code
My Code：
```
public void playSound(String soundName) {
        if (isSound == false)
            return;
        getSound(soundName).play();
    }

 public Sound getSound(String soundName) {
        if (assets.isLoaded(soundName, Sound.class) == true) {
            return assets.get(soundName, Sound.class);
        } else {
            assets.load(soundName, Sound.class);
            assets.finishLoading();
            return assets.get(soundName, Sound.class);
        }
    }

```

#### Version of LibGDX and/or relevant dependencies
Libgdx:1.9.8
Idea:2017.2

#### Please select the affected platforms
- [x] Android

**Comment by whitecostume** [comment]
I looked up the data and found the android sound problem.
We can use the OnLoadCompleteListener to avoid this problem.
This is My Code:
public class AndroidSoundPool extends SoundPool {

    private HashMap<Integer,SoundInfo> cacheMap = new HashMap<Integer,SoundInfo>();
    /**
     * @param maxStreams
     * @param streamType
     * @param srcQuality
     * @deprecated
     */
    public AndroidSoundPool(int maxStreams, int streamType, int srcQuality) {
        super(maxStreams, streamType, srcQuality);

        setOnLoadCompleteListener(new OnLoadCompleteListener() {
            @Override
            public void onLoadComplete(SoundPool soundPool, int i, int i1) {
                if (cacheMap.containsKey(i)){
                    SoundInfo soundInfo = cacheMap.get(i);
                    play(soundInfo.soundID,soundInfo.leftVolume,soundInfo.rightVolume,soundInfo.priority,soundInfo.loop,soundInfo.rate);
                    cacheMap.remove(i);
                }
            }
        });
    }

    public void clearCache(){
        cacheMap.clear();
    }

    public  int playSound(int soundID, float leftVolume, float rightVolume, int priority, int loop, float rate) {
        int streamId =   play( soundID, leftVolume, rightVolume, priority, loop, rate);

        if (streamId == 0){
            cacheMap.put(soundID,new SoundInfo(soundID, leftVolume, rightVolume, priority, loop, rate));
        }

        return streamId;
    }

    class SoundInfo {
        int soundID;float leftVolume;float rightVolume;int priority;int loop;float rate;
        SoundInfo(int soundID, float leftVolume, float rightVolume, int priority, int loop, float rate){
            this.soundID = soundID;
            this.leftVolume = leftVolume;
            this.rightVolume = rightVolume;
            this.priority = priority;
            this.loop = loop;
            this.rate = rate;
        }
    }

}
Instead of SoundPool.After the test, it works. What do you think?

**Comment by mgsx-dev** [comment]
IMO, the main problem with your solution is the synchronization issue : you don't really know exactly when sounds will be played which could cause a lot of sounds to be played at the same time and then audio overload/saturations which could be brutal for player's ears.

I see 2 ways to properly solve it : 

* since all libgdx assets creation are blocking (and properly handled in asset loaders), we could make the andoird sound loading blocking as well to guaranteeing that sound is loaded and usable after constructor call.
* add a loaded state for sounds (all backends). this state could be used for some games when sound loading is critical : waiting for required sounds to be loaded before starting.

I would go for the first one but maybe the community have further recommendations on this subjet though.

Note : it would be great to link your PR to this issue by using smart commit message.

**Comment by Tom-Ski** [comment]
Unfortunately this isn't really something that can be solved cleanly. There are a lot of issues related to this (which you can see on this tracker, and the android tracker).  

The best way to avoid this problem is to simply not load on demand, load the sounds up front and then play after loading phase, not straight after.

If we try to block, other assets can be blocked entirely for huge amounts of time as android sound pool loading can get locked.

Loading state is probably the nicest way to do this, and leave the implementation up to each dev. Playing onComplete will destroy your ears if you are playing multiple sounds like this as often SoundPool seems to loads in batches.

**Comment by whitecostume** [comment]
II know  load resource can solve this problem, but because it is a music type game, if there is too much resource to load and advance load, I hope there are other ways to solve it

**Comment by MrStahlfelge** [comment]
As this is a longstanding wontfix issue, I close this now.
